drivers/sensor/grow_r502a/grow_r502a.c

```diff
@@ -0,0 +1,753 @@
+/*
+ * Copyright (c) 2021 Linumiz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT hzgrow_r502a
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#include "grow_r502a.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(GROW_R502A, CONFIG_SENSOR_LOG_LEVEL);
+
+static void uart_receive_data(const struct device *dev)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+
+	drv_data->buf_len = (size_t)ring_buf_size_get(&drv_data->ringbuf);
+	ring_buf_get(&drv_data->ringbuf, drv_data->buffer, drv_data->buf_len);
+
+	LOG_HEXDUMP_DBG(drv_data->buffer, drv_data->buf_len, "RX");
+	k_sem_give(&drv_data->uart_rx_sem);
+}
+
+static void send_packet(const struct device *dev, uint8_t pkg_type, uint8_t *pkg_data, uint16_t len)
+{
+	const struct grow_r502a_config *cfg = dev->config;
+	struct grow_r502a_data *drv_data = dev->data;
+
+	uint8_t buffer[64] = {0};
+	uint16_t check_sum, pkg_len;
+
+	pkg_len = len + FPS_CHECKSUM_LEN;
+	check_sum = pkg_len + pkg_type;
+
+	sys_put_be16(FPS_STARTCODE, &drv_data->buffer[FPS_STARTCODE_IDX]);
+	sys_put_be32(cfg->comm_addr, &drv_data->buffer[FPS_ADDRESS_IDX]);
+	drv_data->buffer[FPS_PID_IDX] = pkg_type;
+	sys_put_be16(pkg_len, &drv_data->buffer[FPS_PKG_LEN_IDX]);
+	for (int i = FPS_HEADER_LEN; i < (len + FPS_HEADER_LEN); i++) {
+		drv_data->buffer[i] = pkg_data[i - FPS_HEADER_LEN];
```

Comment:
all the buffer to send_packet api are internal to the driver, avoid copying

Reply:
Okay, this too addressed in the latest commits